File: netlify/functions/llm_service.py
from groq import Groq
import logging
from config import Config

logger = logging.getLogger(__name__)

class LLMService:
    def __init__(self):
        self.config = Config()
        self.client = Groq(api_key=self.config.GROQ_API_KEY)
    
    def generate_response(self, query: str, context: str) -> str:
        """Generate response using Groq LLM"""
        system_prompt = f"""
You are Suresh's AI Assistant, representing AI developer and consultant Suresh Beekhani.

CONTEXT:
{context}

INSTRUCTIONS:
- Use the context to answer questions about Suresh's work, projects, and services.
- If context is missing or irrelevant, give general but accurate AI/ML insights.
- Be professional, approachable, and concise.
- Always stay in character as Suresh's AI Assistant.

FORMAT:
Answer: [clear response]
Context Used: [summary or "No relevant context"]
Next Steps: [optional suggestions]
"""

        try:
            completion = self.client.chat.completions.create(
                messages=[
                    {"role": "system", "content": system_prompt},
                    {"role": "user", "content": query}
                ],
                model=self.config.LLM_MODEL,
                temperature=self.config.TEMPERATURE,
                max_tokens=self.config.MAX_TOKENS
            )
            
            return completion.choices[0].message.content or "I encountered an issue generating a response. Please try again."
            
        except Exception as e:
            logger.error(f"Error generating response with Groq: {e}")
            return "I'm having trouble generating a response right now. Please try again shortly."

# Global instance
llm_service = LLMService()

File: netlify/functions/llm-service.js
import Groq from 'groq-sdk';
import { config } from './config.js';
import { RetryHandler } from './retry-handler.js';

class LLMService {
  constructor() {
    this.client = null;
    if (config.GROQ_API_KEY) {
      this.client = new Groq({
        apiKey: config.GROQ_API_KEY,
      });
    }
  }

  async generateResponse(query, context = '') {
    if (!this.client) {
      console.error('Groq client not initialized - missing API key');
      throw new Error('API key not configured');
    }

    console.log('Generating response for query:', query.substring(0, 50) + '...');
    console.log('Context provided:', context ? 'Yes' : 'No');

    const systemPrompt = `
I'm Suresh Beekhani's Personal AI Assistant. I provide clear, confident guidance on AI, ML, Generative AI, NLP, Computer Vision, Predictive Analytics, and Chatbots across industries like HealthTech, FinTech, EdTech, GreenTech, Retail, and E-Commerce.

Notable projects include AI-Driven Law GPT, predictive diagnostics, intelligent chatbots, real-time fraud detection, and computer vision solutions.

FORMATTING: Use plain text only - no markdown, no asterisks, no bullet points. Write naturally like a human conversation.

${context ? `\n\nCONTEXT:\n${context}\n\nUse this context for accurate answers.` : ''}
`;

    try {
      const response = await this.generateResponseWithRetry(systemPrompt, query);
      console.log('✅ Generated response successfully');
      return response;
    } catch (error) {
      console.error('❌ Error generating response:', error.message);
      throw error;
    }
  }

  async generateResponseWithRetry(systemPrompt, query) {
    const operation = async () => {
      const chatCompletion = await this.client.chat.completions.create({
        messages: [
          { role: "user", content: systemPrompt + "\n\nUser: " + query }
        ],
        model: "deepseek-r1-distill-llama-70b",
        temperature: 0.6,
        max_completion_tokens: 4096,
        top_p: 0.95,
        stream: true,
        stop: null
      });

      let fullResponse = '';
      for await (const chunk of chatCompletion) {
        const content = chunk.choices[0]?.delta?.content || '';
        fullResponse += content;
      }

      // Remove <think> tags
      fullResponse = fullResponse.replace(/<think>[\s\S]*?<\/think>/g, '').trim();
      fullResponse = fullResponse.replace(/<think>[\s\S]*$/g, '').trim();

      if (!fullResponse || fullResponse.length < 10) {
        throw new Error('Empty or too short response from model');
      }

      return fullResponse;
    };

    return await RetryHandler.withRetry(operation, {
      maxRetries: 2,
      baseDelay: 500,
      maxDelay: 2000,
      timeout: 60000,
      onRetry: (attempt, error) => {
        console.log(`🔄 Retrying (attempt ${attempt}): ${error.message}`);
      }
    });
  }
}

export const llmService = new LLMService();
